### PR TITLE
Restore dict based typesafe caching

### PR DESCRIPTION
**Pull request by ethanhs**
This also restores `isfile_case` caching and fixes a type error (and legitimate bug) (!).

The last remaining follow up from #4693 is to run only one Python subprocess.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Does this need to go into release-0.590? I.e. is it a critical bugfix to the PEP 561 code or just a performance improvement? (I haven't looked yet -- was there a perf regression in situations that don't invoke PEP 561?)

File: mypy/stubgen.py [hunk 1]
```diff
@@ -160,7 +160,8 @@ def find_module_path_and_all(module: str, pyversion: Tuple[int, int],
             module_all = getattr(mod, '__all__', None)
     else:
         # Find module by going through search path.
-        module_path = mypy.build.FindModuleCache().find_module(module, ['.'] + search_path,
+        module_path = mypy.build.FindModuleCache().find_module(module,
+                                                               ('.',) + tuple(search_path),
```
Comment (ethanhs) [inline mypy/stubgen.py]: @gvanrossum this is a bug on master. `lib_path` should be a tuple, and as you can see to the left, it is passed as a list. I can split this fix out if you don't want the rest.

Comment (JelleZijlstra) [reply mypy/stubgen.py]: What is the concrete effect here? Could this cause problems to users?

Comment (ethanhs) [reply mypy/stubgen.py]: Essentially, `lru_cache` won't be able to hash the passed `list`.

```
$ python -m mypy.stubgen --no-import mypy
Traceback (most recent call last):
  File "C:\Python36\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Python36\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\ethanhs\Documents\mypy\mypy\stubgen.py", line 1022, in <module>
    main()
  File "C:\Users\ethanhs\Documents\mypy\mypy\stubgen.py", line 901, in main
    raise e
  File "C:\Users\ethanhs\Documents\mypy\mypy\stubgen.py", line 898, in main
    include_private=options.include_private)
  File "C:\Users\ethanhs\Documents\mypy\mypy\stubgen.py", line 106, in generate_stub_for_module
    interpreter=interpreter)
  File "C:\Users\ethanhs\Documents\mypy\mypy\stubgen.py", line 164, in find_module_path_and_all
    interpreter)
TypeError: unhashable type: 'list'
```

Comment (JelleZijlstra) [reply mypy/stubgen.py]: Then maybe make a separate PR for that change?

Comment (gvanrossum) [reply mypy/stubgen.py]: OK, please split that out -- I can cherry-pick a 3-line fix to stubgen with a lot less risk and effort than this whole PR.

Comment (ethanhs) [reply mypy/stubgen.py]: @gvanrossum https://github.com/python/mypy/pull/4899

**Pull request merged**